src/chttpd/src/chttpd_db.erl

```diff
@@ -825,21 +826,135 @@ multi_all_docs_view(Req, Db, OP, Queries) ->
-    {ok, Resp} = fabric2_db:fold_docs(Db, fun view_cb/2, VAcc, Options),
-    {ok, Resp#vacc.resp}.
+    VAcc0 = #vacc{
+        db = Db,
+        req = Req,
+        threshold = Max
+    },
+    case Args1#mrargs.keys of
+        undefined ->
+            Options = [
+                {user_ctx, Req#httpd.user_ctx},
+                {dir, Args1#mrargs.direction},
+                {start_key, Args1#mrargs.start_key},
+                {end_key, Args1#mrargs.end_key},
+                {limit, Args1#mrargs.limit},
+                {skip, Args1#mrargs.skip}
+            ],
+            Acc = {iter, Db, Args1, VAcc0},
+            {ok, {iter, _, _, Resp}} =
+                    fabric2_db:fold_docs(Db, fun view_cb/2, Acc, Options),
+            {ok, Resp#vacc.resp};
+        Keys0 when is_list(Keys0) ->
+            Keys1 = case Args1#mrargs.direction of
+                fwd -> Keys0;
+                _ -> lists:reverse(Keys0)
+            end,
+            Keys2 = case Args1#mrargs.skip < length(Keys1) of
+                true -> lists:nthtail(Args1#mrargs.skip, Keys1);
+                false -> []
+            end,
+            Keys3 = case Args1#mrargs.limit < length(Keys2) of
+                true -> lists:sublist(Keys2, Args1#mrargs.limit);
+                false -> Keys2
+            end,
+            %% namespace can be _set_ to `undefined`, so we
+            %% want simulate enum here
+            Namespace = case couch_util:get_value(namespace, Args1#mrargs.extra) of
+                <<"_all_docs">> -> <<"_all_docs">>;
+                <<"_design">> -> <<"_design">>;
+                <<"_local">> -> <<"_local">>;
+                _ -> <<"_all_docs">>
+            end,
+            TotalRows = fabric2_db:get_doc_count(Db, Namespace),
+            Meta = case Args1#mrargs.update_seq of
+                false ->
+                    [{total, TotalRows}, {offset, null}];
+                true ->
+                    [{total, TotalRows}, {offset, null}, {update_seq, null}]
+            end,
+            {ok, VAcc1} = view_cb({meta, Meta}, VAcc0),
+            DocOpts = case Args1#mrargs.conflicts of
+                true -> [conflicts | Args1#mrargs.doc_options];
+                _ -> Args1#mrargs.doc_options
+            end ++ [{user_ctx, Req#httpd.user_ctx}],
+            VAcc2 = lists:foldl(fun(DocId, Acc) ->
+                OpenOpts = [deleted | DocOpts],
+                {Row0, Doc} = case fabric2_db:open_doc(Db, DocId, OpenOpts) of
+                    {not_found, missing} ->
+                        {#view_row{key = DocId}, undefined};
+                    {ok, #doc{deleted = true, revs = Revs}} ->
+                        {RevPos, [RevId | _]} = Revs,
+                        Value = {[
+                            {rev, couch_doc:rev_to_str({RevPos, RevId})},
+                            {deleted, true}
+                        ]},
+                        {#view_row{key=DocId, id=DocId, value=Value}, null};
```

Comment:
Noticed here we return `null` for Doc but if not found we return `undefined`. Is that intentional for some logic downstream or could we return just one of those?

Reply:
undefined means that the `doc` key doesn't show up in the output. `null` means you get `"doc": null` in the output which signifies that its been deleted and isn't just missing. Its a fairly silly distinction IMO but is an API observable thing that we should keep consistent.